### ISSUE TITLE
Need option to disable copying per multi-match in MailMover

### ISSUE THREAD

**Issue by amerlyq** [afewmail/afew#242]
Copying mails matching multiple queries is supposed to be killer feature, but it creates too much frustration for me. In my setup I manage every action-item in `=INBOX` and expect auto-archiving of full threads, when I remove `unread` and `flagged` tags from ALL mails in these threads.
This is what I have configured in notmuch:
```
$ notmuch config list
...
maildir.synchronize_flags=true
index.header.List=List-Id
index.header.To=To
query.abandoned=folder:INBOX AND NOT query:important
query.concluded=query:abandoned AND query:sentreply
query.idontcare=query:abandoned AND NOT query:sentreply
query.important=thread:"{folder:INBOX AND (tag:unread OR tag:flagged)}"
query.sentreply=thread:"{folder:INBOX AND tag:replied}"
```
And that's how I'm forced to use it in `afew`:
```ini
INBOX =
  'NOT tag:flagged AND tag:log':log
  'NOT tag:flagged AND NOT tag:log AND tag:spam':spam
  'query:abandoned AND NOT (tag:log OR tag:spam) AND thread:"{folder:INBOX AND tag:my}"':my
  'query:abandoned AND NOT (tag:log OR tag:spam OR thread:"{folder:INBOX AND tag:my}") AND tag:info':info
  'query:abandoned AND NOT (tag:log OR tag:spam OR thread:"{folder:INBOX AND tag:my}" OR tag:info) AND thread:"{folder:INBOX AND tag:org}"':org
  'query:abandoned AND NOT (tag:log OR tag:spam OR thread:"{folder:INBOX AND tag:my}" OR tag:info OR thread:"{folder:INBOX AND tag:org}") AND tag:junk':junk
  'query:abandoned AND NOT (tag:junk OR thread:"{folder:INBOX AND tag:my}" OR tag:info OR thread:"{folder:INBOX AND tag:org}") AND query:sentreply':work
  'query:abandoned AND NOT (tag:junk OR thread:"{folder:INBOX AND tag:my}" OR tag:info OR thread:"{folder:INBOX AND tag:org}" OR query:sentreply)':dontcare

log =
  'tag:build':log.build
  'tag:maint':log.maint
  'tag:report':log.report
  'from:jira':log.jira
  'from:confluence':log.confl
  'from:stash':log.stash
```
As you can see -- it's a hellish labour to ensure all tags are independent.
I want option to disable copying messages into two matching locations -- and apply rules line by line.
So I could simplify query to simply:
```ini
INBOX =
  'tag:flagged OR NOT query:abandoned':INBOX
  'tag:log':log
  'tag:spam':spam
  'thread:"{folder:INBOX AND tag:my}"':my
  'tag:info':info
  'thread:"{folder:INBOX AND tag:org}"':org
  'tag:junk':junk
  'query:sentreply':work
  '*':dontcare
```
To implement this, `afew` must *shadow* select emails to automatically prevent considering them for next query automatically. Or simply execute each line one by one and immediately move messages and update notmuch db -- to prevent next line matching already moved messages.

Alternative variant, to accomodate both worlds -- copying and not -- expects to extend config format -- so queries inside single assignments will copy, but next assignments will be executed only after full DB sync.
I.e. first query allows copying, but next one will work with clean state after first completes.
```ini
INBOX =
  'tag:log':log
  'tag:spam':spam

INBOX = 'thread:"{folder:INBOX AND tag:my}"':my
INBOX = 'tag:info':info
INBOX = 'thread:"{folder:INBOX AND tag:org}"':org
INBOX = 'tag:junk':junk
INBOX = 'query:sentreply':work
INBOX = '*':dontcare

```
It's a pity, that you can't have same key multiple times in config.

What can we do to achieve desired option?

**Comment by GuillaumeSeren** [comment]
Hey,

Yes I use it mostly to move, according to the tag of a message (unread, spam, archived) and to sync its state to the imap server, its been in a refactoring discussion for some time now, so ideas and patches are welcome.

> As you can see -- it's a hellish labour to ensure all tags are independent.
> I want option to disable copying messages into two matching locations -- and apply rules line by line.

You can do that by using more precise query, as you are showing at the begin at the beginning,
I use a similar logic on my side:
```
MaxHpf/INBOX = 'tag:spam':MaxHpf/spam 'tag:sent':MaxHpf/sent
MaxHpf/archive = 'tag:':MaxHpf/INBOX 'tag:spam':MaxHpf/spam 'tag:sent':MaxHpf/sent
MaxHpf/spam = 'NOT tag:spam':MaxHpf/INBOX 'tag:sent':MaxHpf/sent
MaxHpf/sent = 'NOT tag:sent':MaxHpf/INBOX 'tag:spam':MaxHpf/spam
```

> Or simply execute each line one by one and immediately move messages and update notmuch db -- to prevent next line matching already moved messages.

This is actually the behavior of the proposal I've made for the mailmover(https://github.com/afewmail/afew/pull/189), feel
free to test it but it is the same idea move one bye one and update db.

**Comment by amerlyq** [comment]
> You can do that by using more precise query
> I use a similar logic on my side
Exactly, still your example still lacks a little to showcase the scale of the problem :)

I had seen #189 but it isn't obvious from description if it allows to accomplish what I thought for.
I will trust your words. Hey, @flokli , @andir -- can we restart integration work for #189 -- after all more than year had passed from the time its changes were ~almost~ approved?

**Comment by amerlyq** [comment]
I see some huge activity in commits history.
Does any of newly introduced changes address the issue at least laterally?

**Comment by flokli** [comment]
Recent commits went into cleaning things up (dropping py2 support, implementing some best practices), and adding some tests for some aspects of MailMover.

I think the PR needs a rebase on latest master, and some more tests.
